src/connection/host_lookup.c

```diff
@@ -0,0 +1,435 @@
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+int
+gethostbyname_r_uhost (const char *name, struct hostent *ret, struct hostent_data *ht_data)
+#else
+struct hostent *
+gethostbyname (const char *name)
+#error "HAVE_GETHOSTBYNAME_R"
+#endif				/* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif				/* HAVE_GETHOSTBYNAME_R */
+{
+
+  struct hostent *hp_memcpy = NULL;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+//err also modified by callee
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+      return gethostbyname_r (name, ret, buf, buflen, result, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+      return gethostbyname_r (name, ret, buf, buflen, h_errnop);
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+      return gethostbyname_r (name, ret, &ht_data);
+#else
+      return gethostbyname (name);
+#error "HAVE_GETHOSTBYNAME_R"
+#endif
+#endif /* !HAVE_GETHOSTBYNAME_R */
+    }
+
+  hp_memcpy = host_lookup_internal (name, NULL, HOSTNAME_TO_IPADDR);
+  memcpy ((void *) ret, (void *) hp_memcpy, sizeof (struct hostent));
+
+#ifdef HAVE_GETHOSTBYNAME_R
+#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+  (*result) = (struct hostent *) malloc (sizeof (struct hostent));
+  memcpy ((void *) *result, (void *) hp_memcpy, sizeof (struct hostent));
+  return 0;
+#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return ret;
+#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
+  if (ret == NULL)
+    {
+      return EFAULT;
+    }
+
+  return 0;
+#else
+#error "HAVE_GETHOSTBYNAME_R"
+#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
+#endif /* HAVE_GETHOSTBYNAME_R */
+}
+
+int
+getnameinfo_uhost (struct sockaddr *addr, socklen_t addrlen, char *host, size_t hostlen, char *serv, size_t servlen,
+		   int flags)
+{
+  struct hostent *hp;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return getnameinfo (addr, addrlen, host, hostlen, NULL, 0, NI_NOFQDN);
+    }
+
+  hp = host_lookup_internal (NULL, addr, IPADDR_TO_HOSTNAME);
+
+  if (hp == NULL)
+    {
+      return EAI_NODATA;
+    }
+
+  strcpy (host, hp->h_name);
+  return 0;
+}
+
+int
+getaddrinfo_uhost (char *node, char *service, struct addrinfo *hints, struct addrinfo **res)
+{
+
+  struct hostent *hp = NULL;
+  struct addrinfo results_out;
+  struct sockaddr_in addr_convert;
+  struct in_addr *in_addr_buf;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return getaddrinfo (node, NULL, hints, res);
+    }
+
+  memset (&results_out, 0, sizeof (results_out));
+
+  in_addr_buf = (struct in_addr *) malloc (sizeof (struct in_addr));
```

Comment:
Do you need malloc?
If necessary, it is better to do it after line 407.
And, there is no code in case malloc fails.

Reply:
1) malloc is needed to convert type easy. 


```  
memcpy (&in_addr_buf->s_addr, hp->h_addr_list[0], sizeof (in_addr_buf->s_addr));
memcpy (&addr_convert.sin_addr, &in_addr_buf->s_addr, sizeof (addr_convert.sin_addr));
```

these parts are so bothering so, I choose to in_addr_buf as pointer. and then malloc is needed.

2) Right. I will move it to line 407 and remove malloc fails.